### ISSUE TITLE
Annotation view reuse causes annotation views to disappear when the backing annotation coordinate changes to be located out of the viewport

### ISSUE THREAD

**Issue by bulbuliator** [mapbox/mapbox-gl-native#8489]
**Platform:** iOS
**Mapbox SDK version:** 3.5.0-rc.2

### Steps to trigger behavior

 1. Add view-based annotation as described in [example](https://www.mapbox.com/ios-sdk/examples/annotation-views/)
 2. Change coordinate randomly on 200-300 meters by Core Animation UIView.animate method
 3. Zoom to level >= 13

### Expected behavior

When new coordinate is assigned out of the screen bounds move annotation animated to the screen edge 

### Actual behavior

Annotation disappears instantly without animation

**Comment by boundsj** [comment]
Hi @bulbuliator. I don't think I understand from the repro steps above what the exact issue is or what you are trying to accomplish. Can you please provide a code sample that clearly illustrates what you are trying to do? You may also find the [demo application's implicit animation implementation](https://github.com/mapbox/mapbox-gl-native/blob/ios-v3.4.2/platform/ios/app/MBXAnnotationView.m#L52-L63) useful. Thanks!

**Comment by bulbuliator** [comment]
Hi, @boundsj. 

<details> 
  <summary>My code sample </summary>
   
```swift
class ViewController: UIViewController, MGLMapViewDelegate {

    var point = MGLPointAnnotation()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let mapView = MGLMapView(frame: view.bounds)
        mapView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        mapView.styleURL = MGLStyle.darkStyleURL(withVersion: 9)
        mapView.tintColor = .lightGray
        mapView.centerCoordinate = CLLocationCoordinate2D(latitude: 0, longitude: 66)
        mapView.zoomLevel = 3
        mapView.delegate = self
        view.addSubview(mapView)
        
        let coordinate = CLLocationCoordinate2D(latitude: 0, longitude: 66)
        point.coordinate = coordinate
        point.title = "\(coordinate.latitude), \(coordinate.longitude)"
        
        mapView.addAnnotation(point)
        
        startLocationSimulatorTimer()
        
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        stopLocationSimulatorTimer()
    }
    
    // MARK: - Random annotation moving methods
    
    // Timer to simulate point moving
    var locationSimulatorTimer:Timer?
    
    // Start location simulator timer
    func startLocationSimulatorTimer() {
        
        self.locationSimulatorTimer?.invalidate()
        self.locationSimulatorTimer = Timer.scheduledTimer(timeInterval: 2, target: self, selector: #selector(self.moveAnnotationRandomly), userInfo: nil, repeats: true)
        
    }
    
    // Stop location simulator timer
    func stopLocationSimulatorTimer() {
        self.locationSimulatorTimer?.invalidate()
    }
    
    func moveAnnotationRandomly() {
        
        let newRandomCoordinate = getRandomCoordinate(fromCoordinate: point.coordinate, distanceRadius: 20)
        UIView.animate(withDuration: 1.0) {
            self.point.coordinate = newRandomCoordinate
        }
        
    }
    
    // Return coordinate generated randomly
    // by parameter fromCoordinate and random distanceRadius value
    func getRandomCoordinate(fromCoordinate: CLLocationCoordinate2D, distanceRadius: Double) -> CLLocationCoordinate2D {
        
        let latitudeDelta = drand48() * distanceRadius
        let longitudeDelta = drand48() * distanceRadius
        
        var sign = 1.0
        if Int(arc4random_uniform(UInt32(2))) == 0 {
            sign = -1.0
        }
        
        var randomCoordinate = CLLocationCoordinate2D()
        randomCoordinate.latitude = fromCoordinate.latitude + (latitudeDelta * sign)
        randomCoordinate.longitude = fromCoordinate.longitude + (longitudeDelta * sign)
        
        return randomCoordinate
        
    }
    
    
    
    // MARK: - MGLMapViewDelegate methods
    
    // This delegate method is where you tell the map to load a view for a specific annotation. To load a static MGLAnnotationImage, you would use `-mapView:imageForAnnotation:`.
    func mapView(_ mapView: MGLMapView, viewFor annotation: MGLAnnotation) -> MGLAnnotationView? {
        // This example is only concerned with point annotations.
        guard annotation is MGLPointAnnotation else {
            return nil
        }
        
        // Use the point annotation’s longitude value (as a string) as the reuse identifier for its view.
        let reuseIdentifier = "\(annotation.coordinate.longitude)"
        
        // For better performance, always try to reuse existing annotations.
        var annotationView = mapView.dequeueReusableAnnotationView(withIdentifier: reuseIdentifier)
        
        // If there’s no reusable annotation view available, initialize a new one.
        if annotationView == nil {
            annotationView = CustomAnnotationView(reuseIdentifier: reuseIdentifier)
            annotationView!.frame = CGRect(x: 0, y: 0, width: 40, height: 40)
            
            annotationView!.backgroundColor = UIColor.green
        }
        
        return annotationView
    }
    
    func mapView(_ mapView: MGLMapView, annotationCanShowCallout annotation: MGLAnnotation) -> Bool {
        return true
    }

}

// MGLAnnotationView subclass
class CustomAnnotationView: MGLAnnotationView {
    override func layoutSubviews() {
        super.layoutSubviews()
        
        // Force the annotation view to maintain a constant size when the map is tilted.
        scalesWithViewingDistance = false
        
        // Use CALayer’s corner radius to turn this view into a circle.
        layer.cornerRadius = frame.width / 2
        layer.borderWidth = 2
        layer.borderColor = UIColor.white.cgColor
    }
    
    override func setSelected(_ selected: Bool, animated: Bool) {
        super.setSelected(selected, animated: animated)
        
        // Animate the border width in/out, creating an iris effect.
        let animation = CABasicAnimation(keyPath: "borderWidth")
        animation.duration = 0.1
        layer.borderWidth = selected ? frame.width / 4 : 2
        layer.add(animation, forKey: "borderWidth")
    }
}
```
</details>

**Comment by boundsj** [comment]
Hi @bulbuliator. Thanks so much for the excellent code sample! We really appreciate the help 😃 

I've updated the title of this ticket to reflect the underlying issue. The problem is that, when a coordinate is set to be outside of the viewport, the view reuse mechanism engages and effectively hides the annotation view (it actually moves the view far away from the current viewport so, even with long animations, the annotation view's presentation layer is moved out of the viewport too quickly).

I think we could fully support the `UIView.animate(withDuration:, animations:)` set of APIs by adding a guard to avoid annotation view reuse if the annotation view's layer still has objects in its `animationKeys` at the time that view reuse would otherwise occur. In fact, we may want to consider avoiding any change to the annotation model in `updateAnnotationViews` if that model is associated with a view that is currently being animated.

cc @jmkiley

**Comment by 1ec5** [comment]
Fixed in #8565.
